### PR TITLE
Vet clinic Milestone 4 : Add join visits

### PR DESCRIPTION
**Pull request by Caren-Koroeny**
<h1> Hello Reviewer(s) </h1>

> The above updates were made:

1. Add specialization table
2. Add visits table
3. update animals table to include the primary key
4. Add vets table
5. Add records to vets, specialization and visits table
6. Add queries to perform operations on tables with many-to-many relationship

<h2>Screenshot of query results</h2>
<h3>Who was the last animal seen by William Tatcher?

How many different animals did Stephanie Mendez see?
</h3>

Screenshot from 2023-04-06 18-57-03

<h3>
List all vets and their specialties, including vets with no specialties.

List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020.</h3>

Screenshot from 2023-04-06 18-58-51

<h3>
What animal has the most visits to vets?

Who was Maisy Smith's first visit?</h3>

Screenshot from 2023-04-06 19-03-47

<h3> Details for most recent visit: animal information, vet information, and date of visit.

How many visits were with a vet that did not specialize in that animal's species?</h3>

Screenshot from 2023-04-06 19-04-51

<h3> What specialty should Maisy Smith consider getting? Look for the species she gets the most.</h3>

Screenshot from 2023-04-06 19-06-14

<h2> Happy Reviewing ....</h2>

### CODE REVIEW

File: data.sql [hunk 1]
```diff
@@ -94,4 +88,39 @@ UPDATE animals SET owner_id = 5 WHERE name LIKE 'Angemon';
...
+('Maisy Smith', 26, 'Jan 17, 2019'),
+('Stephanie Mendez', 64, 'May 4, 1981'),
+('Jack Harkness', 38, 'Jun 8, 2008');
+
+-- insert data into the specializations table
+INSERT INTO specializations (vets, species) VALUES (1, 1),
+(3, 1),
+(3, 2),
+(4, 2);
+
```
Comment (bobb-Rob) [inline data.sql:105]: - [ ] [OPTIONAL] This insertion could have been done dynamically where you do not know the `vet_id` and `specie_id`. 


```sql
 INSERT INTO specializations (species_id, vet_id)
 VALUES ((SELECT id FROM vets WHERE name = 'William Tatcher'), (SELECT id FROM species WHERE name = 'Pokemon'));

 INSERT INTO specializations (species_id, vet_id)
 VALUES ((SELECT id FROM vets WHERE name = 'Stephanie Mendez'), (SELECT id FROM species WHERE name = 'Pokemon'));

 INSERT INTO specializations (species_id, vet_id)
 VALUES ((SELECT id FROM vets WHERE name = 'Stephanie Mendez'), (SELECT id FROM species WHERE name = 'Digimon'));

 INSERT INTO specializations (species_id, vet_id)
 VALUES ((SELECT id FROM vets WHERE name = 'Jack Harkness'), (SELECT id FROM species WHERE name = 'Digimon'));
```

File: data.sql [hunk 2]
```diff
@@ -94,4 +88,39 @@ UPDATE animals SET owner_id = 5 WHERE name LIKE 'Angemon';
...
+(6,2, 'Apr 7, 2021'),
+(7,4, 'Sep 29, 2019'),
+(8,3, 'Oct 3, 2020'),
+(8,3, 'Nov 4, 2020'),
+(9,2, 'Jan 24, 2019'),
+(9,2, 'May 15, 2019'),
+(9,2, 'Feb 27, 2020'),
+(9,2, 'Aug 3, 2020'),
+(10,3, 'May 24, 2020'),
+(10,1, 'Jan 11, 2021');
```
Comment (bobb-Rob) [inline data.sql:126]: - [ ] [OPTIONAL] Similarly here, your method works fine but it is easy to make mistakes using your method. Using dynamic query you are less likely to make an error.  see example below 👇🏾 

```sql
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Agumon'), (SELECT id FROM vets WHERE name = 'William Tatcher'), '2020-05-24');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Agumon'), (SELECT id FROM vets WHERE name = 'Stephanie Mendez'), '2020-07-22');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Gabumon'), (SELECT id FROM vets WHERE name = 'Jack Harkness'), '2021-02-02');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Pikachu'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-01-05');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Pikachu'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-03-08');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Pikachu'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-05-14');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Devimon'), (SELECT id FROM vets WHERE name = 'Stephanie Mendez'), '2021-05-04');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Charmander'), (SELECT id FROM vets WHERE name = 'Jack Harkness'), '2021-02-24');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Plantmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2019-12-21');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Plantmon'), (SELECT id FROM vets WHERE name = 'William Tatcher'), '2020-08-10');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Plantmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2021-04-07');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Squirtle'), (SELECT id FROM vets WHERE name = 'Stephanie Mendez'), '2019-09-29');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Angemon'), (SELECT id FROM vets WHERE name = 'Jack Harkness'), '2020-10-03');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Angemon'), (SELECT id FROM vets WHERE name = 'Jack Harkness'), '2020-11-04');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Boarmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2019-01-24');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Boarmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2019-05-15');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Boarmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-02-27');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Boarmon'), (SELECT id FROM vets WHERE name = 'Maisy Smith'), '2020-08-03');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Blossom'), (SELECT id FROM vets WHERE name = 'Stephanie Mendez'), '2020-05-24');
 INSERT INTO visits (animal_id, vet_id, vist_date) VALUES ((SELECT id FROM animal WHERE name = 'Blossom'), (SELECT id FROM vets WHERE name = 'William Tatcher'), '2021-01-11');
```

File: data.sql [hunk 3]
```diff
@@ -1,34 +1,28 @@
...
-   ('Pikachu', '2021-01-07', 1, false, 15.04),
-   ('Devimon', '2017-05-12', 5, true, 11);
+INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
+VALUES ('Agumon', 'Feb 3, 2020', 0, true, 10.23);
+INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
+VALUES ('Gabumon', 'Nov 15, 2018',2 ,true, 8);
+INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
+VALUES ('Pikachu', 'Jan 7, 2021',1 ,false, 15.04);
+INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
+VALUES ('Devimon', 'May 12, 2017',5 ,true, 11);
```
Comment (bobb-Rob) [inline data.sql:11]: - [ ] It appears you did not add in your **_queries.sql file_** queries to answer the following:

    - Who was the last animal seen by William Tatcher?
    - How many different animals did Stephanie Mendez see?
    - List all vets and their specialties, including vets with no specialties.
    - List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020.
    - What animal has the most visits to vets?
    - Who was Maisy Smith's first visit?
    - Details for most recent visit: animal information, vet information, and date of visit.
    - How many visits were with a vet that did not specialize in that animal's species?
    - What specialty should Maisy Smith consider getting? Look for the species she gets the most.

But you provided the screenshot of the results of your queries. Please add these queries to your queries.sql file.

File: data.sql [hunk 4]
```diff
@@ -1,34 +1,28 @@
...
 
--- update table
```
Comment (bobb-Rob) [inline data.sql:9]: - [ ] Using your screenshots, this query - **_List all vets and their specialties, including vets with no specialties._**,

Some vets appear more than once and at other times it looks like they do not have a specialty. E.g William Tatcher's specialty is `Pokemon`, but his name appears for the third time without a specialty. I believe this is not expected. 

Please try the query below and see if the result is different. 

```sql
SELECT vets.*, species.name
  FROM vets
  LEFT JOIN specializations ON vets.id = specializations.vet_id
  LEFT JOIN species ON specializations.species_id = species.id;
```

File: data.sql [hunk 5]
```diff
@@ -1,34 +1,28 @@
...
 INSERT INTO animals (name, date_of_birth, escape_attempts, neutered, weight_kg)
```
Comment (bobb-Rob) [inline data.sql:12]: - [ ] For the following queries 
    - _Who was the last animal seen by William Tatcher?_
    - _How many different animals did Stephanie Mendez see?_
    - _List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020._ 
    -  _What animal has the most visits to vets?_
    - _Who was Maisy Smith's first visit?_
    - _Details for most recent visit: animal information, vet information, and date of visit._
    - _How many times visits where with a vet that did not specialize in that animal's species?_
    -  _What specialty should Maisy Smith consider getting? Look for the species she gets the most._

The screenshot showed the correct value expected. Well done 👏🏾

**Review by bobb-Rob — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️
Hi @Caren-Koroeny, 

Good job so far! 👏👏👏 
There are some issues that you still need to work on to go to the next project but you are almost there!  😃 
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>


## Highlights
✅ No Linter error
✅ Correct use of Gitflow
✅ Very descriptive Pull Request(PR)


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: data.sql (hunk 1 above)
Comment (Caren-Koroeny) [reply data.sql]: I made the changes, thank you sir

File: data.sql (hunk 4 above)
Comment (Caren-Koroeny) [reply data.sql:9]: I did it again and this is what am now getting.

Screenshot from 2023-04-06 20-50-00

**Review by BenMukebo — APPROVED** [review]
### Hi @Caren-Koroeny, 👋 

# STATUS: Approved  🎉 

Great work on making the changes requested by a previous reviewer :1st_place_medal:  
 Your project is complete!  There is nothing else to say other than... it's time to merge it
Congratulations! 🎉  🎉 🎉 



### To Highlight :1st_place_medal: 
- Your Pull Request has a proper title and description. ✔️  


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me on( @BenMukebo ) in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**